arbitrary-file-write/file-delete: Add FreeBSD specific

lsalvadore opened the pull request:
FreeBSD does not have /etc/shadow. So I have tested /etc/master.passwd, /etc/pwd.db and /etc/spwd.db.

I have created a FreeBSD subsection in the Test Notes section. I am not sure this is the best way to do it: when the repository grows we will probably have resources specific to different operating systems, different linux distributions, different linux kernels etc. I fear this will become messy quickly.

Two different ideas to fix the issue:
- use different directories and files for different software and versions, as you have already suggested in https://github.com/LiveOverflow/security-research/pull/1#issuecomment-1201008257 . It has the advantage that you can do everything with .md files, immediately readable on the GitHub website;
- be format agnostic and focus on data by using xml: this would give us more flexibility and we could change our mind on the repository structure more easily as it grows; moreover, this would also allow us to organize our data in multiple format if we wanted (e.g. we could create tools to generate  an md version, an asciiDoc version, an html version, a pdf version etc.)

Please let me know if I have to fix anything in my pull request.

lsalvadore committed:
arbitrary-file-write/file-delete: Add FreeBSD specific

FreeBSD does not have /etc/shadow. Test /etc/master.passwd, /etc/pwd.db
and /etc/spwd.db instead.

Tested OS:	FreeBSD 13.1-RELEASE

LiveOverflow commented:
Thanks for checking! I actually realized we totally forgot to test reboots. What if these password, sudo and other auth related files are missing on a reboot. In a local-privilege escalation scenario this could be useful. I will soon check this for Ubuntu, but it would be awesome if we could have that for FreeBSD as well!

I think a .xml, .json or yaml format will make sense at some point. But I think we should collect more examples to make sure we invent a format that fits all cases.

lsalvadore committed:
arbitrary-file-write/file-delete: Add FreeBSD test

Test what the consequences of removing /etc/spwd.db, /etc/master.passwd,
/etc/pwd.db are after the machine is rebooted.

Tested OS:	FreeBSD 13.1-RELEASE

lsalvadore committed:
arbitrary-file-write/file-delete: Add FreeBSD test

Test removing /etc/group.

Tested OS:	FreeBSD 13.1-RELEASE

lsalvadore committed:
arbitrary-file-write/file-delete: Add FreeBSD test

- Define ${LOCALBASE} variable.
- Test removing ${LOCALBASE}/etc/sudoers.

Tested OS:	FreeBSD 13.1-RELEASE

lsalvadore commented:
I have tested the consequences of the files removals after rebooting and committed the results. I have also tested what happens when `/etc/group` and the `sudoers` files are removed on FreeBSD (reboot test included).

In the case of `/etc/group`, of course some things must break (I have already spotted some errors about unknown groups in the boot messages), but for now I have only written that the tools that you tested on Linux work fine on FreeBSD: although we are searching for things that break, I think it is interesting that those things do not break on this particular OS. Feel free to ask me to remove the line about `/etc/group` if you disagree.

Also, since we are not tracking versions in the repository files yet, I have added this information in the commit messages ("Tested OS: FreeBSD 13.1-RELEASE"), so it is easier to add it in the future if we need it. Feel free to ask me to add more information if you think it might be useful (I don't think it is important to note that I made my test on a virtual machine in this case, is it?).
